src/anyio/_backends/_asyncio.py

```diff
@@ -663,38 +696,43 @@ async def __aexit__(
         exc_val: BaseException | None,
         exc_tb: TracebackType | None,
     ) -> bool | None:
-        ignore_exception = self.cancel_scope.__exit__(exc_type, exc_val, exc_tb)
         if exc_val is not None:
             self.cancel_scope.cancel()
             if not isinstance(exc_val, CancelledError):
                 self._exceptions.append(exc_val)
 
-        cancelled_exc_while_waiting_tasks: CancelledError | None = None
-        while self._tasks:
-            try:
-                await asyncio.wait(self._tasks)
-            except CancelledError as exc:
-                # This task was cancelled natively; reraise the CancelledError later
-                # unless this task was already interrupted by another exception
-                self.cancel_scope.cancel()
-                if cancelled_exc_while_waiting_tasks is None:
-                    cancelled_exc_while_waiting_tasks = exc
+        try:
+            if self._tasks:
+                with CancelScope() as wait_scope:
+                    while self._tasks:
+                        try:
+                            await asyncio.wait(self._tasks)
+                        except CancelledError as exc:
+                            # Shield the scope against further cancellation attempts,
+                            # as they're not productive (#695)
+                            wait_scope.shield = True
+                            self.cancel_scope.cancel()
+                            if exc_val is None:
+                                exc_val = exc
```

Comment:
I identified one uncancellation issue: that the cancellation counter is incremented for every `task.cancel()` - regardless of the task to be cancelled, and then subsequently used to cancel the *host task*. The counter should only be incremented when the *host task* is cancelled. I've added the fix and a test that failed without that fix.

Reply:
Now that I think about this more, I believe the behavior is still incorrect if there are nested task groups with their own host tasks. :face_exhaling:

Reply:
These two tests will not both pass without some deeper changes:
```py
    async def test_cancel_counter_nested_scopes(self) -> None:
        with CancelScope() as root_scope:
            with CancelScope():
                root_scope.cancel()
                await checkpoint()

        assert not cast(asyncio.Task, asyncio.current_task()).cancelling()

    async def test_uncancel_after_taskgroup_cancelled(self) -> None:
        """
        Test that a cancel scope only uncancels the host task as many times as it has
        cancelled that specific task, and won't count child task cancellations towards
        that amount.

        """
        async def child_task(task_status: TaskStatus[None]) -> None:
            async with create_task_group() as tg:
                tg.start_soon(sleep, 3)
                await wait_all_tasks_blocked()
                task_status.started()

        task = asyncio.current_task()
        assert task
        with pytest.raises(CancelledError):
            async with create_task_group() as tg:
                await tg.start(child_task)
                task.cancel()

        assert task.cancelling() == 1
```

Reply:
I believe that the correct fix would be to always increment the specific scope's cancel counter when the scope cancels its own host task, even when the cancellation originates from an enclosing scope. And then, on exit, each scope would always decrement its own counter and uncancel its host task just as many times.

Reply:
I'm wondering if there's hope for restoring 3.8 compatibility after these changes.

Reply:
My fixes are not working; maybe I need to re-read @gschaffner 's code and see if that helps.

Reply:
After a good night's sleep I had another look and realized that a single line change would fix both tests.

Reply:
@gschaffner Ok, so here is my 2 cents on your first test: why should the task group swallow cancellations it didn't cause? That flies in the face of the whole idea of the uncancellation system. If the task is cancelled twice, it should exit with a `CancelledError` and `task.cancelling() == 2` because it was externally cancelled twice.

Reply:
As for the second test, that's a bit trickier. Because it's being cancelled first by the cancel scope, the first raised `CancelledError` is the one raised for the cancel scope, and the next ones are then ignored. To fix this, I think the task group should prioritize native cancellations over any cancel scope-originated ones.

Reply:
I've adapted the first test, fixing the expected `cancelling()` count, and it passes now. Hopefully this was it!

Reply:
Regarding your `test_asyncio_external_cancellation_swallowed_during_timeout()`, you're explicitly cancelling a task but the code responsible that is not uncancelling it when it's being caught, which then messes up the cancel counter.

My point is that we if we catch N cancellation errors, we can totally swallow N-1 of them so long as we make sure to reraise one of them. Imagine if we had two asyncio timeouts:
```py
async with asyncio.timeout(5), asyncio.timeout(3), create_task_group() as tg:
    tg.start_soon(refuses_cancellation_for_10_seconds)
```
This will cause the task group to catch two native cancellation errors, but it will only reraise one. When the task group is exited, I expected `task.cancelling()` to return 2 – one for each `asyncio.timeout()`. Then, each timeout CM will decrement the cancel counter which then goes to zero once they're both exited.

As for your last snippet, why do you expect `task.cancelling()` to be nonzero in the `finally:` block? The outer scope did not need to cancel the task, as the effective scope was already explicitly cancelled. As there are no yield points between exiting the inner scope and exiting the outer scope, the outer scope does not have a chance (or need to) to actually cancel the task.

Case in point – this passes:
```py
    async def test_cancelled_scope_based_checkpoint(self) -> None:
        """Regression test closely related to #698."""
        task = asyncio.current_task()
        assert task
        with CancelScope() as outer_scope:
            outer_scope.cancel()

            try:
                # The following three lines are a way to implement a checkpoint
                # function. See also https://github.com/python-trio/trio/issues/860.
                with CancelScope() as inner_scope:
                    inner_scope.cancel()
                    await checkpoint()
            finally:
                try:
                    await checkpoint()
                finally:
                    assert task.cancelling() == 1

            pytest.fail("checkpoint should have raised")

        assert not task.cancelling()
```

Reply:
Do you think the docs are wrong here? Or do you think that I have just misunderstood them.

I think the docs are correct about the cancel counter, but it doesn't really apply to us since we're not really swallowing any native cancellations anywhere because we always reraise _something_ when we catch a `CancelledError`. Only if we didn't do that would we need to think about the cancel counters.

Reply:
That `test_uncancel_after_child_task_failed()` case does look like it should not leak any cancellations, and I should fix it.

Reply:
Ok, fixed via 90812136f5032085de7f335cea0406d1b4e515ea.

Reply:
In `test_uncancel_laurel_like()`, the user code uncancels a cancellation it did not initiate, and this is a bug. Cancellation and uncancellation must always go hand in hand. This in fact was a problem in `CancelScope` implementation before my latest fix.

Reply:
This puts the stdlib docs in questionable light indeed:
> If end-user code is, for some reason, suppressing cancellation by catching [CancelledError](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError), it needs to call this method to remove the cancellation state.

I feel that this might, after all, be misleading at best. I will need to take a look at the implementation of `asyncio.timeout()` to gain a better understanding of how it's used in the stdlib.